sw/device/lib/base/csr.h

```diff
@@ -0,0 +1,505 @@
+    case kCsrMhpmcounter22:
+      return "mhpmcounter22";
+    case kCsrMhpmcounter23:
+      return "mhpmcounter23";
+    case kCsrMhpmcounter24:
+      return "mhpmcounter24";
+    case kCsrMhpmcounter25:
+      return "mhpmcounter25";
+    case kCsrMhpmcounter26:
+      return "mhpmcounter26";
+    case kCsrMhpmcounter27:
+      return "mhpmcounter27";
+    case kCsrMhpmcounter28:
+      return "mhpmcounter28";
+    case kCsrMhpmcounter29:
+      return "mhpmcounter29";
+    case kCsrMhpmcounter30:
+      return "mhpmcounter30";
+    case kCsrMhpmcounter31:
+      return "mhpmcounter31";
+    case kCsrMcycleh:
+      return "mcycleh";
+    case kCsrMinstreth:
+      return "minstreth";
+    case kCsrMhpmcounter3h:
+      return "mhpmcounter3h";
+    case kCsrMhpmcounter4h:
+      return "mhpmcounter4h";
+    case kCsrMhpmcounter5h:
+      return "mhpmcounter5h";
+    case kCsrMhpmcounter6h:
+      return "mhpmcounter6h";
+    case kCsrMhpmcounter7h:
+      return "mhpmcounter7h";
+    case kCsrMhpmcounter8h:
+      return "mhpmcounter8h";
+    case kCsrMhpmcounter9h:
+      return "mhpmcounter9h";
+    case kCsrMhpmcounter10h:
+      return "mhpmcounter10h";
+    case kCsrMhpmcounter11h:
+      return "mhpmcounter11h";
+    case kCsrMhpmcounter12h:
+      return "mhpmcounter12h";
+    case kCsrMhpmcounter13h:
+      return "mhpmcounter13h";
+    case kCsrMhpmcounter14h:
+      return "mhpmcounter14h";
+    case kCsrMhpmcounter15h:
+      return "mhpmcounter15h";
+    case kCsrMhpmcounter16h:
+      return "mhpmcounter16h";
+    case kCsrMhpmcounter17h:
+      return "mhpmcounter17h";
+    case kCsrMhpmcounter18h:
+      return "mhpmcounter18h";
+    case kCsrMhpmcounter19h:
+      return "mhpmcounter19h";
+    case kCsrMhpmcounter20h:
+      return "mhpmcounter20h";
+    case kCsrMhpmcounter21h:
+      return "mhpmcounter21h";
+    case kCsrMhpmcounter22h:
+      return "mhpmcounter22h";
+    case kCsrMhpmcounter23h:
+      return "mhpmcounter23h";
+    case kCsrMhpmcounter24h:
+      return "mhpmcounter24h";
+    case kCsrMhpmcounter25h:
+      return "mhpmcounter25h";
+    case kCsrMhpmcounter26h:
+      return "mhpmcounter26h";
+    case kCsrMhpmcounter27h:
+      return "mhpmcounter27h";
+    case kCsrMhpmcounter28h:
+      return "mhpmcounter28h";
+    case kCsrMhpmcounter29h:
+      return "mhpmcounter29h";
+    case kCsrMhpmcounter30h:
+      return "mhpmcounter30h";
+    case kCsrMhpmcounter31h:
+      return "mhpmcounter31h";
+    case kCsrMhartid:
+      return "mhartid";
+    default:
+      return "unknown";
+  }
+}
+
+uint32_t mock_csr_read(csr_t csr);
+uint32_t mock_csr_write(csr_t csr, uint32_t value);
+uint32_t mock_csr_set_bits(csr_t csr, uint32_t mask);
+uint32_t mock_csr_clear_bits(csr_t csr, uint32_t mask);
+
+#endif  // MOCK_CSR
+
+__attribute__((always_inline)) inline uint32_t csr_read(csr_t csr) {
+#ifdef MOCK_CSR
+  // Apply the same immediate constraint as the real code.
+  // asm volatile("" :: "i"(csr));
```

Comment:
@luismarques Compiling the unit tests with this line uncommented leads to `error: asm operand 0 probably doesn’t match constraints [-Werror]`. I think this is because they are compiled without optimizations so the csr variable isn't optimized away and therefore isn't a constant expression resulting in the immediate constraint failing. I haven't dug into it though. Anyway, do let me know if you have any observations.

Reply:
Long story short, AFAIK there's nothing that can be done about this. Your current solution with the macros really is that way to go for C code, unfortunately.